main/lsp/requests/workspace_symbols.cc

```diff
@@ -21,12 +44,134 @@ unique_ptr<SymbolInformation> symbolRef2SymbolInformation(const LSPConfiguration
+                queryMatchLength++;
+                if (gap == 0) {
+                    // no penalty for adjacent matching characters
+                } else if (atBoundary) {
+                    score += 100 + gap; // being *on* a boundary is good
+                } else {
+                    score += 200 + 5 * gap; // being *off* a boundary is bad
+                }
+                break;
+            }
+            atBoundary = !isalnum(symbolCh);
+            gap++;
+        }
+        if (queryMatchLength > 1) {
+            score += symbolCharsConsumed;
+        }
+        if (symbolIter == symbolEnd) {
+            break;
+        }
+    }
+    return make_pair(score, queryMatchLength);
+}
+} // namespace
+
+vector<unique_ptr<SymbolInformation>> SymbolMatcher::doQuery(string_view query_view, int limit) {
+    vector<unique_ptr<SymbolInformation>> results;
+    string query(query_view);
+    if (query.empty()) {
+        return results;
+    }
+    const int queryLength = query.length();
+
+    struct ScoreInfo {
+        u4 symbolIndex = 0;
+        int lengthMatched = 0;
+        int score = 0;
+    };
+    vector<ScoreInfo> scoreInfos(gs.symbolsUsed());
+    // First pass: match short names
+    for (u4 symbolIndex = 1; symbolIndex < gs.symbolsUsed(); symbolIndex++) {
+        auto symbolData = core::SymbolRef(gs, symbolIndex).data(gs);
+        auto nameData = symbolData->name.data(gs);
+        if (nameData->kind == core::NameKind::UNIQUE) {
+            continue;
+        }
+        auto shortName = nameData->shortName(gs);
+        auto [score, lengthMatched] = scoreSymbol(shortName, query);
+        scoreInfos[symbolIndex].symbolIndex = symbolIndex;
+        scoreInfos[symbolIndex].score = score;
+        scoreInfos[symbolIndex].lengthMatched = lengthMatched;
+    }
+
+    // Second pass: in some cases, try again if matching long name might improve the result
+    for (u4 symbolIndex = 1; symbolIndex < gs.symbolsUsed(); symbolIndex++) {
+        ScoreInfo &scoreInfo = scoreInfos[symbolIndex];
+        if (scoreInfo.symbolIndex == 0) {
+            continue; // symbol ineligible
+        }
+        auto symbolData = core::SymbolRef(gs, symbolIndex).data(gs);
+        auto ownerRef = symbolData->owner;
+        if (!ownerRef.exists()) {
+            continue; // no owner
+        }
+        ScoreInfo &ownerScoreInfo = scoreInfos[ownerRef._id];
+        if (ownerScoreInfo.lengthMatched == 0) {
+            continue; // owner can't improve search result
+        }
+        if (ownerScoreInfo.lengthMatched == queryLength) {
+            continue; // don't bother retrying: it fully matches the container
+        }
+        if (ownerScoreInfo.lengthMatched <= scoreInfo.lengthMatched && ownerScoreInfo.score >= scoreInfo.score) {
+            continue; // owner's score was same-or-worse on equal-or-shorter match, don't bother
+        }
+        auto fullName = symbolData->showFullName(gs);
```

Comment:
I expect this single line to dominate runtime of this algorithm.

Reply:
Is it "fast enough"?
If not, I think there's a easy dynamic programming we can do.
Happy to chat IRL to explain how to make it fast.

Reply:
It's totally fast enough, but a power user would likely appreciate a faster version. You two should chat in case users desire faster search in the future.

Reply:
Excellent point...fixing now!